### PR TITLE
use span for scatter/gather lists

### PR DESCRIPTION
**Pull request by ssiloti**

### CODE REVIEW

File: src/disk_buffer_pool.cpp [hunk 1]
```diff
@@ -219,21 +219,22 @@ namespace libtorrent
...
-			if (iov[i].iov_base == nullptr)
+			i.iov_base = allocate_buffer_impl(l, "pending read");
+			i.iov_len = block_size();
+			if (i.iov_base == nullptr)
 			{
 				// uh oh. We failed to allocate the buffer!
 				// we need to roll back and free all the buffers
 				// we've already allocated
-				for (int j = 0; j < i; ++j)
+				for (auto j : iov)
```
Comment (aldenml) [inline src/disk_buffer_pool.cpp]: I know that probably the extra cycles will have zero impact, but I find more explicit the algorithm idea as it was, even if it's trivial. Just an opinion.

File: src/file.cpp [hunk 2]
```diff
@@ -336,11 +336,11 @@ static_assert((libtorrent::file::sparse & libtorrent::file::attribute_mask) == 0
 
 namespace libtorrent
 {
-	int bufs_size(file::iovec_t const* bufs, int num_bufs)
+	int bufs_size(span<file::iovec_t const> bufs)
 	{
 		std::size_t size = 0;
-		for (file::iovec_t const* i = bufs, *end(bufs + num_bufs); i < end; ++i)
-			size += i->iov_len;
+		for (auto buf : bufs)
```
Comment (aldenml) [inline src/file.cpp]: I see that sometimes you use `auto` and others `auto&`, any reason not to always use `auto&` (or even `auto const&`)?

File: src/storage.cpp [hunk 3]
```diff
@@ -225,18 +221,15 @@ namespace libtorrent
...
-			, int const size
-			, file::iovec_t const* bufs, storage_error& ec)
+			, span<file::iovec_t const> bufs, storage_error& ec)
 			final
 		{
 			if (m_storage.files().pad_file_at(file_index))
 			{
 				// writing to a pad-file is a no-op
-				return size;
+				return bufs_size(bufs);
```
Comment (aldenml) [inline src/storage.cpp]: was this a bug?

Comment (arvidn) [inline src/storage.cpp]: this function used to take the size, as pre-computed. now it doesn't and have to compute it itself

Comment (arvidn) [inline src/storage.cpp]: I would expect it to be negligible, if any

Comment (aldenml) [inline src/storage.cpp]: ok

File: src/file.cpp (hunk 2 above)
Comment (ssiloti) [inline src/file.cpp]: In general I prefer to handle spans by value. I only use auto& when I need to modify the entries.

File: src/storage.cpp (hunk 3 above)
Comment (ssiloti) [inline src/storage.cpp]: If anything I expect performance to be better. Writing to pad files should happen relatively infrequently compared to writes to regular files and this approach saves an iteration over `bufs` in the regular file case.

File: src/file.cpp (hunk 2 above)
Comment (aldenml) [inline src/file.cpp]: hmm...but I think will end up copying memory if the objects are not trivial in the general case, am I wrong? (not a problem in this case)

Comment (ssiloti) [inline src/file.cpp]: If the objects you're dealing with are large then yes, you would want to use auto const&. I was speaking specifically about spans which are small and trivial to copy so passing by value is preferable.

**Pull request merged**